hack/windows/Install-OVS.ps1

```diff
@@ -244,49 +407,215 @@ function InstallDependency() {
+    if ($LocalFile -ne "") {
+        if (-not (Test-Path $LocalFile)){
+            Log "Path $LocalFile doesn't exist, exit"
+            exit 1
+        }
+
+        $ovsFile = Get-Item $LocalFile
+        if ($ovsFile -is [System.IO.DirectoryInfo])  {
+            return $ovsFile.FullName
+        }
+
+        # $ovsFile as a zip file is supported
+        $attributes = $ovsFile.Attributes
+        if (("$attributes" -eq "Archive") -and ($ovsFile.Extension -eq ".zip" ) ) {
+            $OVSZip = $LocalFile
+            $OVSDownloadURL = ""
+            $OVSPublishedHash = ""
+        } else {
+            Log "Unsupported local file $LocalFile; it should be a zip archive"
+            exit 1
+        }
+    } else {
+        $OVSZip = "$WorkDir\ovs-win64.zip"
+        if ($DownloadURL -ne "" -and $DownloadURL -ne "$OVSDownloadURL") {
+            $OVSDownloadURL = $DownloadURL
+            $desiredOVSPublishedHash = "invalid"
+        }
+    }
+
+    # Extract zip file to $env:TEMP\openvswitch
+    if (Test-Path -Path $tempOVSDir) {
+        rm -r $tempOVSDir
+    }
+    $removeZipFile = $false
+    if ($OVSDownloadURL -ne "") {
+        DownloadOVS -localZipFile $OVSZip -downloadURL $OVSDownloadURL -desiredHash $desiredOVSPublishedHash
+        $removeZipFile = $true
+    }
+    Expand-Archive -Path $OVSZip -DestinationPath $env:TEMP | Out-Null
+    if ($removeZipFile) {
+        rm $OVSZip
+    }
+    return $tempOVSDir
+}
 
-InstallOVS
+function CopyOVSUtilities() {
+    param (
+        [parameter(Mandatory = $true)] [string] $OVSLocalPath
+    )
+    $installUsrBinPath = "${OVSInstallDir}\usr\bin"
+    $usrBinPath="${OVSLocalPath}\usr\bin"
+    Remove-Item -Recurse $installUsrBinPath -ErrorAction SilentlyContinue
+    mkdir -p $installUsrBinPath
+    cp -r $usrBinPath\* $installUsrBinPath
+    AddToEnvPath($installUsrBinPath)
+}
 
-if ($InstallUserspace -eq $true) {
-    InstallDependency
+function InstallOVS() {
+    param (
+        [parameter(Mandatory = $true)] [string] $OVSLocalPath
+    )
+    # Install powershell modules
+    $OVSScriptsPath = "${OVSLocalPath}\scripts"
+    CheckAndInstallScripts($OVSScriptsPath)
+
+    # Install VC redistributables.
+    $OVSRedistDir="${OVSLocalPath}\redist"
+    # Check if the VC redistributable is already installed. If not installed, or the installed version
+    # is lower than $MininalVCRedistVersion, install the local provided VC redistributable files or
+    # download the file and install it.
+    CheckAndInstallVCRedists -VCRedistPath $OVSRedistDir -VCRedistsVersion $MininalVCRedistVersion
+
+    # Install OVS driver.
+    $OVSDriverDir = "${OVSLocalPath}\driver"
+    Log "Installing OVS kernel driver"
+    CheckAndInstallOVSDriver($OVSDriverDir)
+
+    # Install OpenSSL dependencies.
+    $OVSBinPaths="${OVSLocalPath}\usr\bin"
+    if ($InstallUserspace -eq $true) {
+        $OVSBinPaths="${OVSBinPaths};${OVSLocalPath}\usr\sbin"
+    }
+    InstallOpenSSLFiles "$OVSBinPaths"
```

Comment:
it seems that we copy the SSL files to `$OVSBinPaths` (`"${OVSLocalPath}\usr\bin"`), and then we rely on `CopyOVSUtilities` to copy them to the OVS install directory? Was that intentional? I guess it will work but shouldn't `InstallOpenSSLFiles` copy the files to their "final" installation location directly?

Reply:
The current solution is on desired that copy the ssl files to the LocalFile first, then copy the final files to the install dir. This is because the ssl dll files are required by both usr/bin and usr/sbin, and the two directories are copied in different step: usr/bin is copied in step `CopyOVSUtilities`, and usr/sbin is copied in step `InstallOVSServices`. If we plan to copy ssl dll files directly to the final path, the coy operations are distributed into different steps, and we also need to record the path where the dll files are extracted until the end.

Reply:
Ok, sounds good